File: problems/271.cpp
#include <iostream>

using namespace std;

string s, S = "CDEI";
bool solve()
{
    int cnt = 0;
    for (int i = s.size() - 1; i >= 0; --i)
    {
        if (s[i] >= 'p' && s[i] <= 'z')
            ++cnt;
        else if (s[i] == 'N')
        {
            if (cnt == 0)
                return false;
        }
        else if (S.find(s[i]) != string::npos)
        {
            if (cnt < 2)
                return false;

            --cnt;
        }
        else
            return false;
                
    }

    return cnt == 1;
}

int main()
{
    ios::sync_with_stdio(false);
    while (cin >> s)
        cout << (solve()? "YES\n" : "NO\n");
}

File: problems/271.py
from sys import stdin

def solve(s):
    cnt = 0;
    for c in s[::-1]:
        if ord(c) in range(ord('p'), ord('z') + 1):
            cnt += 1
        elif c == 'N':
            if cnt == 0:
                return False
        elif c in "CDEI":
            if cnt < 2:
                return False

            cnt -= 1
        else:
            return False

    return cnt == 1


for s in stdin:
    print("YES" if solve(s[:-1]) else "NO")
